rust/sbp/src/codec.rs

```diff
@@ -0,0 +1,354 @@
+    }
+
+    impl Decoder for SbpCodec {
+        type Item = SBP;
+        type Error = crate::Error;
+
+        fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>> {
+            match parse_sbp(&src) {
+                ParseResult::Ok((bytes_read, msg)) => {
+                    src.advance(bytes_read);
+                    log::trace!("{:?}", msg);
+                    Ok(Some(msg))
+                }
+                ParseResult::Err((bytes_read, err)) => {
+                    src.advance(bytes_read);
+                    log::warn!("{}", err);
+                    Ok(None)
+                }
+                ParseResult::Incomplete => {
+                    src.reserve(crate::MSG_HEADER_LEN + crate::SBP_MAX_PAYLOAD_SIZE);
+                    Ok(None)
+                }
+            }
+        }
+
+        fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>> {
+            match self.decode(buf) {
+                Ok(Some(frame)) => Ok(Some(frame)),
+                _ => Ok(None),
+            }
+        }
+    }
+
+    impl Encoder<SBP> for SbpCodec {
+        type Error = crate::Error;
+
+        fn encode(&mut self, msg: SBP, dst: &mut BytesMut) -> Result<()> {
+            self.frame.clear();
+            match msg.write_frame(&mut self.frame) {
+                Ok(_) => dst.put_slice(self.frame.as_slice()),
+                Err(err) => log::error!("Error converting sbp message to frame: {}", err),
+            }
+            Ok(())
+        }
+    }
+}
+
+#[cfg(feature = "json")]
+pub mod json {
+    use std::collections::HashMap;
+
+    use bytes::{Buf, BufMut, BytesMut};
+    use futures::StreamExt;
+    use serde::{de::DeserializeOwned, Deserialize, Serialize};
+    use serde_json::{Deserializer, Value};
+    use tokio::io::{AsyncRead, AsyncWrite};
+    use tokio_util::codec::{Decoder, Encoder, FramedRead, FramedWrite};
+
+    use crate::{
+        messages::{SBPMessage, SBP},
+        serialize::SbpSerialize,
+        Result,
+    };
+
+    pub async fn json2sbp<R, W>(input: R, output: W) -> Result<()>
+    where
+        R: AsyncRead + Unpin,
+        W: AsyncWrite,
+    {
+        let source = JsonDecoder::decode_reader(input);
+        let sink = super::sbp::SbpCodec::encode_writer(output);
+        source.forward(sink).await
+    }
+
+    pub async fn json2json<R, W>(input: R, output: W) -> Result<()>
+    where
+        R: AsyncRead + Unpin,
+        W: AsyncWrite,
+    {
+        let source = Json2JsonDecoder::decode_reader(input);
+        let sink = JsonEncoder::encode_writer(output);
+        source.forward(sink).await
+    }
+
+    pub async fn sbp2json<R, W>(input: R, output: W) -> Result<()>
+    where
+        R: AsyncRead,
+        W: AsyncWrite,
+    {
+        let source = super::sbp::SbpCodec::decode_reader(input);
+        let sink = JsonEncoder::encode_writer(output);
+        source.forward(sink).await
+    }
+
+    #[derive(Debug, Deserialize)]
+    #[serde(untagged)]
```

Comment:
`untagged` enums de/serialize without the inner names (`Input`/`Nested`) included